src/compute/src/render/continual_task.rs

```diff
@@ -572,35 +573,29 @@ where
     output_stream.as_collection()
 }
 
-// TODO(ct2): Write this as a non-async operator.
 fn step_forward<G, D, R>(name: &str, input: Collection<G, D, R>) -> Collection<G, D, R>
 where
     G: Scope<Timestamp = Timestamp>,
     D: Data,
     R: Semigroup + 'static,
 {
     let name = format!("ct_step_forward({})", name);
-    let mut builder = AsyncOperatorBuilder::new(name.clone(), input.scope());
-    let mut input = builder.new_disconnected_input(&input.inner, Pipeline);
+    let mut builder = OperatorBuilder::new(name.clone(), input.scope());
     let (mut output, output_stream) = builder.new_output();
-    builder.build(move |caps| async move {
-        let [mut cap]: [_; 1] = caps.try_into().expect("one capability per output");
-        loop {
-            let Some(event) = input.next().await else {
-                return;
-            };
-            match event {
-                Event::Data(_data_cap, mut data) => {
-                    for (_, ts, _) in &mut data {
-                        *ts = ts.step_forward();
-                    }
-                    output.give_container(&cap, &mut data);
-                }
-                Event::Progress(progress) => {
-                    if let Some(progress) = progress.into_option() {
-                        cap.downgrade(&progress.step_forward());
-                    }
+    let mut input =
+        builder.new_input_connection(&input.inner, Pipeline, vec![Antichain::from_elem(1.into())]);
+    builder.set_notify(false);
+    builder.build(move |_caps| {
+        let mut buf = Vec::new();
+        move |_frontiers| {
+            let mut output = output.activate();
+            while let Some((cap, data)) = input.next() {
+                data.swap(&mut buf);
+                for (_, ts, _) in &mut buf {
+                    *ts = ts.step_forward();
```

Comment:
This will panic once `ts` reaches the maximum value. Is this a concern? (I think no, but just pointing it out.)

Reply:
Correct, it's fine both at the current state of main (b/c we step back before forward) and also after #30073 (b/c we're using it to advance a `T-1` collection to `T`). Added a mention here in the rustdoc of when it would panic at runtime and I'll also add a comment in #30073 to mention why the usage there is safe.